### PR TITLE
Provide convenient helper for authentication

### PR DESCRIPTION
**Pull request by weppos**
This PR provides convenient helpers and reduces boilerplate.

It adds a dependency to "golang.org/x/oauth2", but we're almost forcing it nonetheless (although indirectly).

In the past I refrained from adding this dependency because oauth2 package required contexts that were not widely adopted yet, and most libraries using in older Go version would have conflicts.

Today we require "contexts" anyways.

Should someone complain about this specific dependency, we can always try to extract the helper in a separate file and use build flags.

### CODE REVIEW

**Comment by aeden** [comment]
given these two helpers are in the `dnsimple` package, is there any reason to *not* return a `dnsimple.Client` instance?

File: README.md [hunk 1]
```diff
@@ -31,9 +31,7 @@ import (
 )
 
 func main() {
-    oauthToken := "xxxxxxx"
-    ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: oauthToken})
-    tc := oauth2.NewClient(context.Background(), ts)
+    tc := dnsimple.StaticTokenClient(context.Background(), "your-token")
```
Comment (aeden) [inline README.md:34]: Given this is an HTTP client, consider including that in the name of the function. `StaticTokenHTTPClient`

File: dnsimple/authentication.go [hunk 2]
```diff
@@ -1,9 +1,24 @@
...
 
 import (
+	"context"
 	"net/http"
+
+	"golang.org/x/oauth2"
 )
 
+// BasicAuthClient returns a client that authenticates via HTTP Basic Auth with given username and password.
+func BasicAuthClient(_ context.Context, username, password string) *http.Client {
```
Comment (aeden) [inline dnsimple/authentication.go:11]: Given this is an HTTP client, consider including that in the name of the function. `BasicAuthHTTPClient`

File: dnsimple/authentication.go [hunk 3]
```diff
@@ -1,9 +1,24 @@
...
+	tp := BasicAuthTransport{Username: username, Password: password}
+	return tp.Client()
+}
+
+// BasicAuthClient returns a client that authenticates with a static OAuth token.
+func StaticTokenClient(ctx context.Context, token string) *http.Client {
```
Comment (aeden) [inline dnsimple/authentication.go:17]: Given this is an HTTP client, consider including that in the name of the function. `StaticTokenHTTPClient`

**Comment by weppos** [comment]
> given these two helpers are in the `dnsimple` package, is there any reason to _not_ return a `dnsimple.Client` instance?

I thought about it, but that will limit the ability to then extend the interface value. For instance, I've seen monitoring tools wrapping a `http.Client` and return another client with instrumentation.

```go
tc := dnsimple.StaticTokenClient(context.Background(), "your-token")
ic := instrumenter.Wrap(tc)
client := dnsimple.NewClient(ic)
```

If we return a dnsimple client, then they won't be able to intercept the HTTP client and wrap it. Since `httpClient` is a widely used interface (and our client happily accept any struct that implements it), I thought it would be better to have the helpers just preparing a HTTP client.

**Comment by jacegu** [comment]
Trying to wrap my head on these changes. Question: what is this _"context"_ you talk about and where does it come from?

**Comment by weppos** [comment]
> Trying to wrap my head around these changes. Question: what is this _"context"_ you talk about and where does it come from?

[`context` is a package](https://golang.org/pkg/context/) introduced in Go 1.7 that allows for [advanced concurrency patterns](https://blog.golang.org/context). It was not part of the initial Go release, therefore several libraries had to introduce significant breaking changes in order to support it, including us https://github.com/dnsimple/dnsimple-go/pull/90.

**Pull request merged**